File: server/src/main/java/bgu/spl/net/impl/stomp/Frame.java
package bgu.spl.net.impl.stomp;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.StringReader;
import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;

public class Frame{
    
    private String command;
    private Map<String, String> headers;
    private String body;

    Frame(String command, Map<String, String> headers, String body){
        this.command = command;
        this.headers = headers;
        this.body = body;
    }

    Frame(String s){
        this("",new HashMap<>(), "");
        BufferedReader reader = new BufferedReader(new StringReader(s));
        
        try {
            command = reader.readLine();
            String line = null;

            while(!(line = reader.readLine()).equals("")){ //parse headers
                int colonIndex = line.indexOf(':');
                String key = line.substring(0, colonIndex);
                String value = line.substring(colonIndex + 1);
                headers.put(key, value);
            }
            
            while((line = reader.readLine()) != null)
                body += line+"\n";
        
        } catch (IOException e) {}
    }

    public String getCommand(){
        return command;
    }

    public Map<String,String> getHeaders(){
        return headers;
    }
    
    public String getBody(){
        return body;
    }
    
    public String toString() { 
        String output = command+"\n";
        if(headers != null){
            Iterator<Map.Entry<String, String>> iterator = headers.entrySet().iterator();
            while (iterator.hasNext()) {
                Map.Entry<String, String> entry = iterator.next();
                String key = entry.getKey();
                String value = entry.getValue();
                output += key+":"+value+"\n";
            }
        }
        output += "\n"+body;
        return output;
     }
}

File: SPL231-Assignment3-student-template-2/client/src/Frame.cpp
#include "../include/Frame.h"



Frame::Frame(string command, map<string, string> &headers, string body) : 
command(command), 
headers(headers), 
body(body){}

Frame::Frame(const string &s) : command(""), headers({}), body("")
{
    vector<string> lines;
    size_t start_line = 0;
    size_t end_line;
    while(end_line != string::npos){
        end_line = s.find("\n", start_line);
        lines.push_back(s.substr(start_line,end_line-start_line));
        start_line = end_line+1;
    }
    
    command = lines[0];
    unsigned i = 1;
    for(; i<lines.size() && lines[i].length() > 0; i++){ //headers
        size_t colon_index = lines[i].find(":");
        string key = lines[i].substr(0, colon_index);
        string value = lines[i].substr(colon_index+1);
        headers.insert(make_pair(key,value));
    }
    for(; i<lines.size()-1 ;i++)
        body += lines[i] + "\n";
    body += lines[i];

}


const string &Frame::get_command() const{
    return command;
}

const map<string, string> &Frame::get_headers() const{
    return headers;
}

const string &Frame::get_body() const{
    return body;
}

const string Frame::to_string() const{
    string output = command+"\n";
    
    for(auto iterator = headers.begin(); iterator != headers.end(); ++iterator){
        output += iterator->first+":"+iterator->second+"\n";
    }
    output += "\n"+body;
    return output;
}
